### PR TITLE
[Model] Support HSDP for LongCat-Image and LongCat-Image-Edit 

### PR DESCRIPTION
**Pull request by MoveCloudROY**
## Purpose
Add HSDP support for LongCat-Image and LongCat-Image-Edit.
This PR:
- Adds HSDP shard conditions for LongCat transformer blocks.
- Fixes LongCat Image Edit prompt encoding device placement under HSDP.
- Adds LongCat online serving smoke cases for `--use-hsdp --hsdp-shard-size 2`.
- Updates the diffusion feature matrix for LongCat HSDP support.
## Test Plan
```bash
python -m pytest tests/diffusion/models/longcat_image/test_longcat_image_hsdp.py tests/diffusion/distributed/test_hsdp.py
python -m pytest --collect-only tests/e2e/online_serving/test_longcat_image_expansion.py tests/e2e/online_serving/test_longcat_image_edit_expansion.py
Manual smoke:
CUDA_VISIBLE_DEVICES=0 python -m vllm_omni.entrypoints.cli.main serve /data0/mllm/LongCat-Image --host 127.0.0.1 --port 8091 --omni
CUDA_VISIBLE_DEVICES=0,1 python -m vllm_omni.entrypoints.cli.main serve /data0/mllm/LongCat-Image --host 127.0.0.1 --port 8091 --omni --use-hsdp --hsdp-shard-size 2
CUDA_VISIBLE_DEVICES=0,1,2,3 python -m vllm_omni.entrypoints.cli.main serve /data0/mllm/LongCat-Image --host 127.0.0.1 --port 8091 --omni --use-hsdp --hsdp-shard-size 4
```
## Test Result

| Config(1344x768 RGB) | Peak Memory per GPU   | Change vs 1 GPU |
| -------------------- | --------------------- | --------------- |
| 1 GPU non-HSDP       | 33310 MiB / 32.53 GiB | baseline        |
| 2 GPU HSDP shard=2   | 29710 MiB / 29.01 GiB | -10.8%          |
| 4 GPU HSDP shard=4   | 26774 MiB / 26.15 GiB | -19.6%          |

Unit tests:
```
tests/diffusion/models/longcat_image/test_longcat_image_hsdp.py .        [  3%]
tests/diffusion/distributed/test_hsdp.py ...........................     [100%]
28 passed, 17 warnings
E2E collection:
collected 6 items
tests/e2e/online_serving/test_longcat_image_expansion.py
  test_longcat_image[single_card_001]
  test_longcat_image[parallel_001]
  test_longcat_image[parallel_hsdp_shard2]
tests/e2e/online_serving/test_longcat_image_edit_expansion.py
  test_longcat_image_edit[single_card_001]
  test_longcat_image_edit[parallel_001]
  test_longcat_image_edit[parallel_hsdp_shard2]
1 GPU non-HSDP LongCat Image smoke:
INFO 05-01 00:39:03 [serve.py:111] Detected diffusion model: /data0/mllm/LongCat-Image
INFO 05-01 00:39:24 [diffusion_worker.py:199] Worker 0: Process-scoped GPU memory after model loading: 27.81 GiB.
(APIServer pid=634538) INFO 05-01 00:39:34 [diffusion_engine.py:127] Generation completed successfully.
Output:
output: 1344x768 RGB
peak memory delta: 33310 MiB / 32.53 GiB
2 GPU HSDP LongCat Image smoke:
INFO 05-01 00:39:51 [serve.py:111] Detected diffusion model: /data0/mllm/LongCat-Image
INFO 05-01 00:40:26 [hsdp.py:128] HSDP Inference: replicate_size=1, shard_size=2, world_size=2, rank=1, fs_world_size=2, fs_rank=1
INFO 05-01 00:40:26 [hsdp.py:128] HSDP Inference: replicate_size=1, shard_size=2, world_size=2, rank=0, fs_world_size=2, fs_rank=0
INFO 05-01 00:40:29 [hsdp.py:202] Sharded 30 modules + root
INFO 05-01 00:40:29 [hsdp.py:173] HSDP applied to model: FSDPLongCatImageTransformer2DModel
INFO 05-01 00:40:30 [diffusion_worker.py:199] Worker 0: Process-scoped GPU memory after model loading: 6.73 GiB.
INFO 05-01 00:40:32 [hsdp.py:202] Sharded 30 modules + root
INFO 05-01 00:40:32 [hsdp.py:173] HSDP applied to model: FSDPLongCatImageTransformer2DModel
INFO 05-01 00:40:33 [diffusion_worker.py:199] Worker 1: Process-scoped GPU memory after model loading: 6.73 GiB.
(APIServer pid=636192) INFO 05-01 00:40:44 [diffusion_engine.py:127] Generation completed successfully.
Output:
output: 1344x768 RGB
peak memory delta: 29710 MiB / 29.01 GiB
change vs 1 GPU: -10.8%
4 GPU HSDP LongCat Image smoke:
INFO 05-01 00:41:01 [serve.py:111] Detected diffusion model: /data0/mllm/LongCat-Image
INFO 05-01 00:41:39 [hsdp.py:128] HSDP Inference: replicate_size=1, shard_size=4, world_size=4, rank=0, fs_world_size=4, fs_rank=0
INFO 05-01 00:41:41 [hsdp.py:128] HSDP Inference: replicate_size=1, shard_size=4, world_size=4, rank=3, fs_world_size=4, fs_rank=3
INFO 05-01 00:41:41 [hsdp.py:128] HSDP Inference: replicate_size=1, shard_size=4, world_size=4, rank=1, fs_world_size=4, fs_rank=1
INFO 05-01 00:41:41 [hsdp.py:128] HSDP Inference: replicate_size=1, shard_size=4, world_size=4, rank=2, fs_world_size=4, fs_rank=2
INFO 05-01 00:41:45 [hsdp.py:202] Sharded 30 modules + root
INFO 05-01 00:41:45 [hsdp.py:173] HSDP applied to model: FSDPLongCatImageTransformer2DModel
INFO 05-01 00:41:46 [hsdp.py:202] Sharded 30 modules + root
INFO 05-01 00:41:46 [hsdp.py:173] HSDP applied to model: FSDPLongCatImageTransformer2DModel
INFO 05-01 00:41:46 [hsdp.py:202] Sharded 30 modules + root
INFO 05-01 00:41:46 [hsdp.py:173] HSDP applied to model: FSDPLongCatImageTransformer2DModel
INFO 05-01 00:41:46 [diffusion_worker.py:199] Worker 1: Process-scoped GPU memory after model loading: 3.84 GiB.
INFO 05-01 00:41:47 [diffusion_worker.py:199] Worker 0: Process-scoped GPU memory after model loading: 3.84 GiB.
INFO 05-01 00:41:47 [diffusion_worker.py:199] Worker 2: Process-scoped GPU memory after model loading: 3.84 GiB.
INFO 05-01 00:41:49 [hsdp.py:202] Sharded 30 modules + root
INFO 05-01 00:41:49 [hsdp.py:173] HSDP applied to model: FSDPLongCatImageTransformer2DModel
INFO 05-01 00:41:50 [diffusion_worker.py:199] Worker 3: Process-scoped GPU memory after model loading: 3.84 GiB.
(APIServer pid=638645) INFO 05-01 00:42:02 [diffusion_engine.py:127] Generation completed successfully.
Output:
output: 1344x768 RGB
peak memory delta: 26774 MiB / 26.15 GiB
change vs 1 GPU: -19.6%
LongCat Image Edit 4 GPU HSDP smoke:
INFO 05-01 00:08:15 [serve.py:111] Detected diffusion model: /data0/mllm/LongCat-Image-Edit
INFO 05-01 00:08:54 [hsdp.py:128] HSDP Inference: replicate_size=1, shard_size=4, world_size=4, rank=3, fs_world_size=4, fs_rank=3
INFO 05-01 00:08:55 [hsdp.py:128] HSDP Inference: replicate_size=1, shard_size=4, world_size=4, rank=1, fs_world_size=4, fs_rank=1
INFO 05-01 00:08:55 [hsdp.py:128] HSDP Inference: replicate_size=1, shard_size=4, world_size=4, rank=2, fs_world_size=4, fs_rank=2
INFO 05-01 00:08:55 [hsdp.py:128] HSDP Inference: replicate_size=1, shard_size=4, world_size=4, rank=0, fs_world_size=4, fs_rank=0
INFO 05-01 00:08:58 [hsdp.py:202] Sharded 30 modules + root
INFO 05-01 00:08:58 [hsdp.py:173] HSDP applied to model: FSDPLongCatImageTransformer2DModel
INFO 05-01 00:08:59 [diffusion_worker.py:199] Worker 2: Process-scoped GPU memory after model loading: 3.84 GiB.
INFO 05-01 00:08:59 [diffusion_worker.py:199] Worker 1: Process-scoped GPU memory after model loading: 3.84 GiB.
INFO 05-01 00:09:00 [diffusion_worker.py:199] Worker 3: Process-scoped GPU memory after model loading: 3.84 GiB.
INFO 05-01 00:09:01 [diffusion_worker.py:199] Worker 0: Process-scoped GPU memory after model loading: 3.84 GiB.
(APIServer pid=618328) INFO 05-01 00:13:53 [diffusion_engine.py:127] Generation completed successfully.
Output:
output: 1024x1024 RGB
```

## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the test style doc (https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [x] (Optional) The necessary documentation update, such as updating supported_models.md and examples for a new model. Please run mkdocs serve to sync the documentation editions to ./docs.
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.

### CODE REVIEW

File: vllm_omni/diffusion/models/longcat_image/pipeline_longcat_image_edit.py [hunk 1]
```diff
@@ -285,6 +285,7 @@ def __init__(
         )
 
     def _encode_prompt(self, prompt, image):
+        text_encoder_device = next(self.text_encoder.parameters()).device
```
[the remark this answers is not in the document]
Comment (MoveCloudROY) [reply vllm_omni/diffusion/models/longcat_image/pipeline_longcat_image_edit.py]: This is intended to resolve the issue where the text_encoder and tensor reside on different devices, but it appears to cause a conflict with CPU offloading.

There are two solutions:
1. Directly check the current offload status via a hook, but this will make the code messy.
2. Move the text_encoder to the device.

I’ve noticed that [a pull request](https://github.com/vllm-project/vllm-omni/pull/2029#pullrequestreview-3988351921) uses method 2, so I will adopt this approach.

File: tests/e2e/online_serving/test_longcat_image_expansion.py [hunk 2]
```diff
@@ -75,4 +89,4 @@ def test_longcat_image(omni_server: OmniServer, openai_client: OpenAIClientHandl
         },
     }
 
-    openai_client.send_diffusion_request(request_config)
+    _ = openai_client.send_diffusion_request(request_config)
```
Comment (RuixiangMa) [inline tests/e2e/online_serving/test_longcat_image_expansion.py]: Unnecessary change in e2e tests

File: tests/e2e/online_serving/test_longcat_image_edit_expansion.py [hunk 3]
```diff
@@ -75,4 +89,4 @@ def test_longcat_image_edit(omni_server: OmniServer, openai_client: OpenAIClient
         },
     }
 
-    openai_client.send_diffusion_request(request_config)
+    _ = openai_client.send_diffusion_request(request_config)
```
Comment (RuixiangMa) [inline tests/e2e/online_serving/test_longcat_image_edit_expansion.py]: Unnecessary change

File: vllm_omni/diffusion/models/longcat_image/pipeline_longcat_image_edit.py [hunk 4]
```diff
@@ -250,6 +250,7 @@ def __init__(
         self.text_encoder = Qwen2_5_VLForConditionalGeneration.from_pretrained(
             model, subfolder="text_encoder", local_files_only=local_files_only
         )
+        self.text_encoder.to(self.device)
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/models/longcat_image/pipeline_longcat_image_edit.py]: merge text_encoder initialization into single line

File: tests/diffusion/models/longcat_image/test_longcat_image_hsdp.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    model = object.__new__(LongCatImageTransformer2DModel)
+    nn.Module.__init__(model)
+    model.transformer_blocks = nn.ModuleList([nn.Linear(4, 4) for _ in range(2)])
+    model.single_transformer_blocks = nn.ModuleList([nn.Linear(4, 4) for _ in range(3)])
+    model.proj_out = nn.Linear(4, 4)
+
+    conditions = cast(
+        list[Callable[[str, nn.Module], bool]],
+        getattr(model, "_hsdp_shard_conditions"),
+    )
```
Comment (RuixiangMa) [inline tests/diffusion/models/longcat_image/test_longcat_image_hsdp.py]: Suggest to use` conditions = getattr(model, "_hsdp_shard_conditions", None) `

File: tests/diffusion/models/longcat_image/test_longcat_image_hsdp.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    condition = conditions[0]
+
+    assert condition("transformer_blocks.0", model.transformer_blocks[0])
+    assert condition("transformer_blocks.1", model.transformer_blocks[1])
+    assert condition("single_transformer_blocks.0", model.single_transformer_blocks[0])
+    assert condition("single_transformer_blocks.1", model.single_transformer_blocks[1])
+    assert condition("single_transformer_blocks.2", model.single_transformer_blocks[2])
+    assert not condition("transformer_blocks", model.transformer_blocks)
+    assert not condition("single_transformer_blocks", model.single_transformer_blocks)
+    assert not condition("proj_out", model.proj_out)
```
Comment (RuixiangMa) [inline tests/diffusion/models/longcat_image/test_longcat_image_hsdp.py]: pls follow the existing pattern — iterate via model.named_modules() + any(cond(...)) , which mirrors the actual consumption path

**Comment by RuixiangMa** [comment]
@MoveCloudROY Thanks for the PR! Left a few comments

**Comment by hsliuustc0106** [comment]
what's the speed perf in each setting?

**Comment by MoveCloudROY** [comment]
@hsliuustc0106 
4xH800
1 warmup + 4 measured，1344x768，4 steps

|                 | 1 GPU        | 2 GPU        | 4 GPU        |
| ------------- | ------------ | ------------ | ------------ |
| No HSDP       | 2.466s       | 2.223s       | 2.224s       |
| HSDP          | 2.645s       | 3.416s       | 4.040s       |
| latency ratio | 1.07x slower | 1.54x slower | 1.82x slower |

**Comment by RuixiangMa** [comment]
LGTM

**Comment by hsliuustc0106** [comment]
Hi @MoveCloudROY, friendly reminder — this PR hasn't had any activity (commits or reviews) in the past 7 days. 🕐

Could you please provide an update?
- If you're still working on it, that's great — just let us know.
- If you're blocked on something, feel free to ask for help.
- If this PR is no longer being pursued, please consider closing it so we can keep the review queue manageable.

Thanks for your contribution! 🙏

**Comment by MoveCloudROY** [comment]
It's all done. Could someone please review and approve it?

**Comment by RuixiangMa** [comment]
@MoveCloudROY Please solve the conflicts.

**Comment by MoveCloudROY** [comment]
Sorry, I accidentally did a force push after rebasing. The conflicts have now been resolved.

**Review by RuixiangMa — APPROVED** [review]
